File: 81_Splat/java/src/Splat.java
import java.util.Arrays;
import java.util.Random;
import java.util.Scanner;

public class Splat {
    private final Random random = new Random();
    private final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        new Splat().run();
    }

    public void run() {
        System.out.printf("%33s%s\n", " ", "SPLAT");
        System.out.printf("%15s%s\n", " ", "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
        System.out.print("\n\n\n");
        System.out.println("WELCOME TO 'SPLAT' -- THE GAME THAT SIMULATES A PARACHUTE");
        System.out.println("JUMP.  TRY TO OPEN YOUR CHUTE AT THE LAST POSSIBLE");
        System.out.println("MOMENT WITHOUT GOING SPLAT.");

        float[] Arr = new float[42];
        Arrays.fill(Arr, 0.0f);
        int K = 0;
        int K1 = 0;
        while (true) {

            System.out.print("\n\n");
            float V = 0.0f;
            float A = 0.0f;
            int N = 0;
            int M = 0;
            int D1 = (int) (9001.0f * random.nextFloat() + 1000);

            System.out.print("SELECT YOUR OWN TERMINAL VELOCITY (YES OR NO) ");
            float V1;
            while (true) {
                String A1 = scanner.next();
                if (A1.equals("NO")) {
                    V1 = (int) (1000 * random.nextFloat());
                    System.out.printf("OK.  TERMINAL VELOCITY = %d MI/HR\n", (int) V1);
                    break;
                }
                if (!A1.equals("YES")) {
                    System.out.print("YES OR NO ");
                    continue;
                }
                System.out.print("WHAT TERMINAL VELOCITY (MI/HR) ");
                V1 = scanner.nextInt();
                break;
            }
            V1 = V1 * (5280.0f / 3600.0f);
            V = V1 + ((V1 * random.nextFloat()) / 20.0f) - ((V1 * random.nextFloat()) / 20.0f);
            System.out.print("WANT TO SELECT ACCELERATION DUE TO GRAVITY (YES OR NO) ");
            float A2;
            while (true) {
                String B1 = scanner.next();
                if (B1.equals("NO")) {
                    A2 = chooseRandomAcceleration();
                    break;
                }
                if (!B1.equals("YES")) {
                    System.out.print("YES OR NO ");
                    continue;
                }
                System.out.print("WHAT ACCELERATION (FT/SEC/SEC) ");
                A2 = scanner.nextFloat();
                break;
            }
            A = A2 + ((A2 * random.nextFloat()) / 20.0f) - ((A2 * random.nextFloat()) / 20.0f);
            System.out.println();
            System.out.printf("    ALTITUDE         = %d FT\n", D1);
            System.out.printf("    TERM. VELOCITY   = %.2f FT/SEC +/-5%%\n", V1);
            System.out.printf("    ACCELERATION     = %.2f FT/SEC/SEC +/-5%%\n", A2);
            System.out.println("SET THE TIMER FOR YOUR FREEFALL.");
            System.out.print("HOW MANY SECONDS ");
            float T = scanner.nextFloat();
            System.out.println("HERE WE GO.\n");
            System.out.println("TIME (SEC)  DIST TO FALL (FT)");
            System.out.println("==========  =================");
            boolean splat = false;
            boolean terminalReached = false;
            float D = 0.0f;
            for (float i = 0.0f; !splat && (i < T); i += T / 8) {
                if (i > (V / A)) {
                    terminalReached = true;
                    System.out.printf("TERMINAL VELOCITY REACHED AT T PLUS %f SECONDS.\n", (V / A));
                    for (i = i; i < T; i += T / 8) {
                        D = D1 - ((V * V / (2 * A)) + (V * (i - (V / A))));
//                    System.out.printf("   ......................................tv %f\n", D);
                        if (D <= 0) {
                            splat = true;
                            break;
                        }
                        System.out.printf("%10.2f  %f\n", i, D);
                    }
                    break;
                }
                D = D1 - ((A / 2) * i * i);
//            System.out.printf("   ......................................debug %f\n", D);
                if (D <= 0) {
                    splat = true;
                    break;
                }
                System.out.printf("%10.2f  %f\n", i, D);
            }

            if (!splat) {

                System.out.println("CHUTE OPEN");
                int J = 0;
                for (J = 0; J < 42; J++) {
                    if (Arr[J] == 0) {
                        Arr[J] = D;
                        break;
                    }
                    K = K + 1;
                    if (D > Arr[J]) {
                        continue;
                    }
                    K1 = K1 + 1;
                }

                if (J > 2) {
                    if (K - K1 <= 0.1 * K) {
                        System.out.printf("WOW!  THAT'S SOME JUMPING.  OF THE %d SUCCESSFUL JUMPS\n", K);
                        System.out.printf("BEFORE YOURS, ONLY %d OPENED THEIR CHUTES LOWER THAN\n", K - K1);
                        System.out.println("YOU DID.");
                    } else if (K - K1 <= 0.25 * K) {
                        System.out.printf("PRETTY GOOD!  %d SUCCESSFUL JUMPS PRECEDED YOURS AND ONLY\n", K);
                        System.out.printf("%d OF THEM GOT LOWER THAN YOU DID BEFORE THEIR CHUTES\n", K - K1);
                        System.out.println("OPENED.");
                    } else if (K - K1 <= 0.5 * K) {
                        System.out.printf("NOT BAD.  THERE HAVE BEEN %d SUCCESSFUL JUMPS BEFORE YOURS.\n", K);
                        System.out.printf("YOU WERE BEATEN OUT BY %d OF THEM.\n", K - K1);
                    } else if (K - K1 <= 0.75 * K) {
                        System.out.printf("CONSERVATIVE, AREN'T YOU?  YOU RANKED ONLY %d IN THE\n", K - K1);
                        System.out.printf("%d SUCCESSFUL JUMPS BEFORE YOURS.", K);
                    } else if (K - K1 <= -0.9 * K) {
                        System.out.println("HUMPH!  DON'T YOU HAVE ANY SPORTING BLOOD?  THERE WERE");
                        System.out.printf("%d SUCCESSFUL JUMPS BEFORE YOURS AND YOU CAME IN %d JUMPS\n", K, K1);
                        System.out.println("BETTER THAN THE WORST.  SHAPE UP!!!\n");
                    } else {
                        System.out.printf("HEY!  YOU PULLED THE RIP CORD MUCH TOO SOON.  %f SUCCESSFUL\n", K);
                        System.out.printf("JUMPS BEFORE YOURS AND YOU CAME IN NUMBER %d!  GET WITH IT!\n", K - K1);
                    }
                } else {
                    System.out.println("AMAZING!!! NOT BAD FOR YOUR ");
                    switch (J) {
                        case 0:
                            System.out.print("1ST ");
                            break;
                        case 1:
                            System.out.print("2ND ");
                            break;
                        case 2:
                            System.out.print("3RD ");
                            break;
                    }
                    System.out.println("SUCCESSFUL JUMP!!!");
                }


            } else {
                if (terminalReached) {
                    System.out.printf("%.2f SPLAT\n", (V / A) + ((D1 - (V * V / (2 * A))) / V));
                } else {
                    System.out.printf("%.2f SPLAT\n", Math.sqrt(2 * D1 / A));

                }
                switch (random.nextInt(10)) {
                    case 0:
                        System.out.println("REQUIESCAT IN PACE.");
                        break;
                    case 1:
                        System.out.println("MAY THE ANGEL OF HEAVEN LEAD YOU INTO PARADISE.");
                        break;
                    case 2:
                        System.out.println("REST IN PEACE.");
                        break;
                    case 3:
                        System.out.println("SON-OF-A-GUN.");
                        break;
                    case 4:
                        System.out.println("#$%&&%!$");
                        break;
                    case 5:
                        System.out.println("A KICK IN THE PANTS IS A BOOST IF YOU'RE HEADED RIGHT.");
                        break;
                    case 6:
                        System.out.println("HMMM. SHOULD HAVE PICKED A SHORTER TIME.");
                        break;
                    case 7:
                        System.out.println("MUTTER. MUTTER. MUTTER.");
                        break;
                    case 8:
                        System.out.println("PUSHING UP DAISIES.");
                        break;
                    default:
                        System.out.println("EASY COME, EASY GO.");

                }
            }
            boolean chosen = false;
            while (!chosen) {
                System.out.print("DO YOU WANT TO PLAY AGAIN ");
                String Z = scanner.next();
                if (Z.equals("YES")) {
                    break;
                }
                if (Z.equals("NO")) {
                    System.out.print("PLEASE ");
                    while (true) {
                        Z = scanner.next();
                        if (Z.equals("YES")) {
                            chosen = true;
                            break;
                        }
                        if (Z.equals("NO")) {
                            System.out.println("SSSSSSSSSS.");
                            return;
                        }
                        System.out.print("YES OR NO ");
                    }
                    continue;
                } else {
                    System.out.print("YES OR NO ");
                }
            }

        }

    }

    private float chooseRandomAcceleration() {
        Planet planet = Planet.pickRandom();
        System.out.printf("%s. ACCELERATION=%.2f FT/SEC/SEC.\n", planet.getMessage(), planet.getAcceleration());
        return planet.getAcceleration();
    }

    enum Planet {
        MERCURY("FINE. YOU'RE ON MERCURY", 12.2f),
        VENUS("ALL RIGHT. YOU'RE ON VENUS", 28.3f),
        EARTH("THEN YOU'RE ON EARTH", 32.16f),
        MOON("FINE. YOU'RE ON THE MOON", 5.15f),
        MARS("ALL RIGHT. YOU'RE ON MARS", 12.5f),
        JUPITER("THEN YOU'RE ON JUPITER", 85.2f),
        SATURN("FINE. YOU'RE ON SATURN", 37.6f),
        URANUS("ALL RIGHT. YOU'RE ON URANUS", 33.8f),
        NEPTUNE("THEN YOU'RE ON NEPTUNE", 39.6f),
        SUN("FINE. YOU'RE ON THE SUN", 896.0f);

        static final Random random = new Random();
        private final String message;
        private final float acceleration;

        Planet(String message, float acceleration) {
            this.message = message;
            this.acceleration = acceleration;
        }

        static Planet pickRandom(){
            return values()[random.nextInt(Planet.values().length)];
        }

        public String getMessage() {
            return message;
        }

        public float getAcceleration() {
            return acceleration;
        }
    }
}

File: 81_Splat/perl/splat.pl
#!/usr/bin/env perl

use 5.010;      # To get 'state' and 'say'

use strict;     # Require explicit declaration of variables
use warnings;   # Enable optional compiler warnings

use English;    # Use more friendly names for Perl's magic variables
use List::Util qw{ shuffle };   # Shuffle an array.
use Scalar::Util qw{ looks_like_number };
use Term::ReadLine;     # Prompt and return user input

our $VERSION = '0.000_01';

use constant ROW_TPLT => ( '%4d' x 8 ) . "\n";

print <<'EOD';
                                 SPLAT
               Creative Computing  Morristown, New Jersey



Welcome to 'Splat' -- the game that simulates a parachute
jump.  Try to open your chute at the last possible 
moment without going splat.
EOD

while ( 1 ) {
    say '';
    my $initial_altitude = int( 9001 * rand() + 1000 );

    my $nominal_terminal_velocity;
    if ( get_yes_no( 'Select your own terminal velocity' ) ) {
        $nominal_terminal_velocity = get_input(
            'What terminal velocity (mi/hr)? ',
            sub { looks_like_number( $ARG ) && $ARG > 0 },
            'Please enter a positive number',
        );
        # Convert miles per hour to feet per second
        $nominal_terminal_velocity = $nominal_terminal_velocity * 5280 / 3600;
    } else {
        $nominal_terminal_velocity = int( 1000 * rand() );
        say "OK.  Terminal velocity = $nominal_terminal_velocity mi/hr"
    }
    my $terminal_velocity = dither( $nominal_terminal_velocity );

    my $nominal_gravity; # Acceleration due to gravity
    if ( get_yes_no( 'Want to select acceleration due to gravity' ) ) {
    } else {
        state $body = [
            [ q<Fine. You're on Mercury. Acceleration = 12.2 ft/sec/sec.>,
                12.2 ],
            [ q<All right. You're on Venus. Acceleration = 28.3 ft/sec/sec.>,
                28.3 ],
            [ q<Then you're on Earth. Acceleration = 32.16 ft/sec/sec.>,
                32.16 ],
            [ q<Fine. You're on the Moon. Acceleration = 5.15 ft/sec/sec.>,
                5.15 ],
            [ q<All right. You're on Mars. Acceleration = 12.5 ft/sec/sec.>,
                12.5 ],
            [ q<Then you're on Jupiter. Acceleration = 85.2 ft/sec/sec.>,
                85.2 ],
            [ q<Fine. You're on Saturn. Acceleration = 37.6 ft/sec/sec.>,
                37.6 ],
            [ q<All right. You're on Uranus. Acceleration = 33.8 ft/sec/sec.>,
                33.8  ],
            [ q<Then you're on Neptune. Acceleration = 39.6 ft/sec/sec.>,
                39.6 ],
            [ q<Fine. You're on the Sun. Acceleration = 896 ft/sec/sec.>,
                896 ],
        ];
        my $pick = $body->[ rand scalar @{ $body } ];
        say $pick->[0];
        $nominal_gravity = $pick->[1];
    }
    my $gravity = dither( $nominal_gravity );

    print <<"EOD";

    Altitude        = $initial_altitude ft
    Term. velocity  = $nominal_terminal_velocity ft/sec +/- 5%
    Acceleration    = $nominal_gravity ft/sec/sec +/- 5%
Set the timer for your freefall
EOD

    my $drop_time = get_input(
        'How many seconds? ',
        sub { m/ \A [0-9]+ \z /smx },
        "Please enter an unsigned integer\n",
    );

    print <<'EOD';
Here we go.

Time (sec)      Dist to fall (ft)
==========      =================
EOD

    if ( defined( my $altitude = make_jump(
                $initial_altitude,
                $gravity,
                $terminal_velocity,
                $drop_time ) )
    ) {
        # Successful jump
        state $succesful = [];
        state $ordinal = [ qw{ 1st 2nd 3rd } ];
        if ( defined( my $ord = $ordinal->[ @{ $succesful } ] ) ) {
            say "Amazing!!! Not nad for your $ord successful jump!!!";
        } else {
            my $jumps = @{ $succesful };
            my $worse = grep { $_ > $altitude } @{ $succesful };
            my $fractile = 1 - $worse / $jumps;
            my $better = $jumps - $worse;
            if ( $fractile <= 0.1 ) {
                print <<"EOD";
Wow!  That's some jumping.  Of the $jumps successful jumps
before yours, only $better opened their chutes lower than 
you did. 
EOD
            } elsif ( $fractile <= 0.25 ) {
                print <<"EOD";
Pretty good! $jumps successful jumps preceded yours and only
$better of them got lower than you did before their chutes
opened.
EOD
            } elsif ( $fractile <= 0.5 ) {
                print <<"EOD";
Not bad.  There have been $jumps successful jumps before yours. 
You were beaten out by $better of them.
EOD
            } elsif ( $fractile <= 0.75 ) {
                print <<"EOD";
Conservative, aren't you?  You ranked only $better in the
$jumps successful jumps before yours.
EOD
            } elsif ( $fractile <= 0.9 ) {
                print <<"EOD";
Humph!  Don't you have any sporting blood?  There were 
$jumps successful jumps before yours and you came in $worse jumps
better than the worst.  Shape up!!!
EOD
            } else {
                print <<"EOD";
Hey!  You pulled the rip cord much too soon.  $jumps successful
jumps before yours and you came in number $better!  Get with it!
EOD
            }
        }
        push @{ $succesful }, $altitude;
    } else {
        # Splat

        say q<I'll give you another chance.>;
    }

    next if get_yes_no( 'Do you want to play again' );
    next if get_yes_no( 'Please' );

    print <<'EOD';
Ssssssssss.

EOD
    last;

}

# Return the first argument modified by up to plus or minus some
# fraction specified by the second argument (default 0.05)
sub dither {
    my ( $arg, $fract ) = @_;
    $fract //= 1 / 20;
    return $arg + ( $arg * rand() * $fract ) - ( $arg * rand() * $fract );
}

use constant FORMAT_FALL    => "%10.1f      %10d\n";
use constant FORMAT_SPLAT   => "%10.1f      %s\n";
sub make_jump {
    my ( $initial_altitude, $gravity, $terminal_velocity, $drop_time ) = @_;
    my $altitude;
    foreach my $step ( 0 .. 8 ) {
        my $time = $step * $drop_time / 8;
        if ( $time > $terminal_velocity / $gravity ) {
            # Terminal velocity reached
            printf "Terminal velocity reached at T plus %.2f seconds.\n",
                $terminal_velocity / $gravity;
            for my $step ( $step .. 8 ) {
                my $time = $step * $drop_time / 8;
                $altitude = $initial_altitude - (
                    $terminal_velocity * $terminal_velocity /
                    ( 2 * $gravity ) + $terminal_velocity * (
                        $time - $terminal_velocity / $gravity ) );
                if ( $altitude > 0 ) {
                    printf FORMAT_FALL, $time, $altitude;
                } else {
                    splat(
                        $terminal_velocity / $gravity + (
                            $initial_altitude -
                            $terminal_velocity * $terminal_velocity /
                            ( 2 * $gravity ) ) / $terminal_velocity,
                    );
                    return;
                }
            }
            last;
        } else {
            $altitude = $initial_altitude - $gravity / 2 * $time * $time;
            if ( $altitude > 0 ) {
                printf FORMAT_FALL, $time, $altitude;
            } else {
                splat( sqrt( 2 * $initial_altitude / $gravity ) );
                return;
            }
        }
    }

    say 'Chute open.';
    return $altitude;
}

sub splat {
    my ( $time ) = @_;
    printf FORMAT_SPLAT, $time, 'Splat!';
    state $rip = [
        q<Requiescat in pace.>,
        q<May the angel of heaven lead you into paradise.>,
        q<Rest in peace.>,
        q<Son-of-a-gun.>,
        q<#$%&&%!$>,
        q<A kick in the pants is a boost if you're headed right.>,
        q<Hmmm. Should have picked a shorter time.>,
        q<Mutter. Mutter. Mutter.>,
        q<Pushing up daisies.>,
        q<Easy come, easy go.>,
    ];
    say $rip->[ rand scalar @{ $rip } ];
    return;
}

# Get input from the user. The arguments are:
# * The prompt
# * A reference to validation code. This code receives the response in
#   $ARG and returns true for a valid response.
# * A warning to print if the response is not valid. This must end in a
#   return.
# The first valid response is returned. An end-of-file terminates the
# script.
sub get_input {
    my ( $prompt, $validate, $warning ) = @ARG;

    # If no validator is passed, default to one that always returns
    # true.
    $validate ||= sub { 1 };

    # Create the readline object. The 'state' causes the variable to be
    # initialized only once, no matter how many times this subroutine is
    # called. The do { ... } is a compound statement used because we
    # need to tweak the created object before we store it.
    state $term = do {
        my $obj = Term::ReadLine->new( 'reverse' );
        $obj->ornaments( 0 );
        $obj;
    };

    while ( 1 ) {   # Iterate indefinitely

        # Read the input into the topic variable, localized to prevent
        # Spooky Action at a Distance. We exit on undef, which signals
        # end-of-file.
        exit unless defined( local $ARG = $term->readline( $prompt ) );

        # Return the input if it is valid.
        return $ARG if $validate->();

        # Issue the warning, and go around the merry-go-round again.
        warn $warning;
    }
}

# Get a yes-or-no answer. The argument is the prompt, which will have
# '? [y/n]: ' appended. The donkey work is done by get_input(), which is
# requested to validate the response as beginning with 'y' or 'n',
# case-insensitive. The return is a true value for 'y' and a false value
# for 'n'.
sub get_yes_no {
    my ( $prompt ) = @ARG;
    state $map_answer = {
        n   => 0,
        y   => 1,
    };
    my $resp = lc get_input(
        "$prompt? [y/n]: ",
        sub { m/ \A [yn] /smxi },
        "Please respond 'y' or 'n'\n",
    );
    return $map_answer->{ substr $resp, 0, 1 };
}

__END__

=head1 TITLE

splat.pl - Play the game 'splat' from Basic Computer Games

=head1 SYNOPSIS

 splat.pl

=head1 DETAILS

This Perl script is a port of C<splat>, which is the 73rd entry in
Basic Computer Games.

This is a very basic port. All I really did was untangle the spaghetti.

=head1 PORTED BY

Thomas R. Wyant, III F<wyant at cpan dot org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2022 by Thomas R. Wyant, III

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl 5.10.0. For more details, see the Artistic
License 1.0 at
L<https://www.perlfoundation.org/artistic-license-10.html>, and/or the
Gnu GPL at L<http://www.gnu.org/licenses/old-licenses/gpl-1.0.txt>.

This program is distributed in the hope that it will be useful, but
without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.

=cut

# ex: set expandtab tabstop=4 textwidth=72 :
